### PR TITLE
[Inliner] Prevent adding pointer attributes to non-pointer arguments

### PR DESCRIPTION
**Pull request by hvdijk**
Fixes a crash seen after #114311

### CODE REVIEW

File: llvm/lib/Transforms/Utils/InlineFunction.cpp [hunk 1]
```diff
@@ -1477,6 +1477,15 @@ static void AddParamAndFnBasicAttributes(const CallBase &CB,
 
         // If so, propagate its access attributes.
         AL = AL.addParamAttributes(Context, I, ValidObjParamAttrs[ArgNo]);
+
+        // If the argument is not a pointer type, remove attributes which only
+        // apply to pointer types.
+        if (!NewInnerCB->getArgOperand(I)->getType()->isPointerTy()) {
+          AL = AL.removeParamAttribute(Context, I, Attribute::ReadNone);
+          AL = AL.removeParamAttribute(Context, I, Attribute::ReadOnly);
```
Comment (goldsteinn) [inline llvm/lib/Transforms/Utils/InlineFunction.cpp:1485]: writeonly too probably.

Comment (hvdijk) [reply llvm/lib/Transforms/Utils/InlineFunction.cpp:1485]: This only needs to handle the attributes that this function may have put in `ValidObjParamAttrs` earlier. `WriteOnly` is not in that list, see https://github.com/llvm/llvm-project/blob/48f6814b2c85f8b519955a9ef604c0e6479b25ba/llvm/lib/Transforms/Utils/InlineFunction.cpp#L1392-L1395 where it only checks these two attributes.

**Review by nikic — CHANGES_REQUESTED** [review]
We should entirely skip the getUnderlyingObject logic for non-pointer arguments instead.

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**